### ISSUE TITLE
Add TriplesReadHandle/TriplesWriteHandle to JSONReadHandle/JSONWriteHandle and XMLReadHandle/XMLWriteHandle

### ISSUE THREAD

**Issue by sammefford** [marklogic/java-client-api#358]
We had no way to use important handles like JacksonHandle and DOMHandle with GraphManager.read/write/merge or SPARQLQueryManager.executeDescribe/executeConstruct.  This is because those all require TriplesReadHandle or TriplesWriteHandle, marker interfaces we failed to associate with JacksonHandle and DOMHandle.  This fix is to add TriplesReadHandle to JSONReadHandle and XMLReadHandle.  Similarly, to add TriplesWriteHandle to JSONWriteHandle and XMLWriteHandle.  That way any JSON or XML handle will work with any method that requires TriplesReadHandle or TriplesWriteHandle.

**Comment by sammefford** [comment]
This is fixed by d4ef824 on develop and cf64fda on 3.0-develop

**Comment by georgeajit** [comment]
Have used JacksonHandle, DOMHandle to verify executeDesribe and executeConstruct methods support for TriplesReadHandle, XMLReadHandle.
GraphManager tests will be handling the write handle's support.

```
/* This test checks a simple SPARQL DESCRIBE results from named graph.
	 * The database should contain triples from rdfjson.json.
	 * Verifies Git Issue 356, 358
	 * 
	 * The DESCRIBE query should be returning one result. Result includes all triples which have the IRI as a subject
     * 
     * Uses Stringhandle and JacksonHandle. 
	 * 
	 */
	@Test
	public void testExecuteDescribeFromRDFJSON() throws IOException, SAXException, ParserConfigurationException
	{	
		System.out.println("In SPARQL Query Manager Test testExecuteDescribeFromRDFJSON method");
		SPARQLQueryManager sparqlQmgr = readclient.newSPARQLQueryManager();
		StringBuffer sparqlQuery = new StringBuffer().append("DESCRIBE <http://example.org/about>");
		
		SPARQLQueryDefinition qdef = sparqlQmgr.newQueryDefinition(sparqlQuery.toString());
		String jsonResults = sparqlQmgr.executeDescribe(qdef, new StringHandle()).get();
     	// Verify result 1 value.
		System.out.println(jsonResults);
		String[] resultString = jsonResults.split(" ");
		assertEquals("Method testExecuteDescribeFromRDFJSON subject is incorrect", "<http://example.org/about>", resultString[0]);
		assertEquals("Method testExecuteDescribeFromRDFJSON predicate is incorrect", "<http://purl.org/dc/elements/1.1/title>", resultString[1]);
		String objectStr = resultString[2] + " " + resultString[3];
		assertEquals("Method testExecuteDescribeFromRDFJSON object is incorrect", "\"Anna's Homepage\"", objectStr);
		
		// Verifying default mime types- Using JacksonHandle
		JacksonHandle jh = new JacksonHandle();
		JsonNode jsonNodes = sparqlQmgr.executeDescribe(qdef, jh).get();
		JsonNode jsonNodeValue = jsonNodes.path("http://example.org/about").path("http://purl.org/dc/elements/1.1/title").get(0);
     	
		// Verify Mime type on the handle and value.
		System.out.println("testQueryOnMultibyeGraphName query result size is " +  jsonNodeValue.get("value"));
		assertEquals("Element person's value incorrect", RDFMimeTypes.RDFJSON, jh.getMimetype());
		assertEquals("Result returned from testQueryOnMultibyeGraphName query is incorrect ", "Anna\'s Homepage", jsonNodeValue.get("value").asText());	
		
		// Verifying default mime types- Using DOMHandle
		DOMHandle dh = new DOMHandle();
		Document xmlDoc = sparqlQmgr.executeDescribe(qdef, dh).get();
		
		Node description = xmlDoc.getFirstChild().getFirstChild();
		NamedNodeMap attrs = description.getFirstChild().getAttributes();
		// attrs NodeMap should have two nodes.
		assertEquals("Attribute length value incorrect", 2, attrs.getLength());
		
		if (attrs.item(0).getNodeName().equalsIgnoreCase("rdf:datatype"))
		{
			assertEquals("Element rdf:datatype value incorrect", "http://www.w3.org/2001/XMLSchema#string", attrs.item(0).getNodeValue());
			assertEquals("Element xmlns data value incorrect", "http://purl.org/dc/elements/1.1/", attrs.item(1).getNodeValue());
		}
		else if (attrs.item(0).getNodeName().equalsIgnoreCase("xmlns"))
		{
			assertEquals("Element rdf:datatype value incorrect", "http://www.w3.org/2001/XMLSchema#string", attrs.item(1).getNodeValue());
			assertEquals("Element xmlns data value incorrect", "http://purl.org/dc/elements/1.1/", attrs.item(0).getNodeValue());
		}
		String value = description.getFirstChild().getFirstChild().getNodeValue();
		assertEquals("Expected value read from DOMHandle incorrect", "Anna\'s Homepage", value);
	}

```

And

```
/* This test checks a SPARQL CONSTRUCT query results.
	 * The database should contain triples from foaf1.nt file. The data format in this file is XML.
	 * Results are returned in a sequence of sem:triple values as triples in memory.
	 * 
	 * The query should be returning one result. This test also verifies Base URI and Git Issue 356, 358.
     * 
     * Uses JacksonHandle and DOMHandle
	 * 
	 */
	@Test
	public void testBaseUri() throws IOException, SAXException, ParserConfigurationException
	{	
		System.out.println("In SPARQL Query Manager Test testBaseUri method");
		SPARQLQueryManager sparqlQmgr = writeclient.newSPARQLQueryManager();

		StringBuffer sparqlQuery = new StringBuffer().append(" PREFIX foaf: <http://xmlns.com/foaf/0.1/>");
		sparqlQuery.append(newline);
		sparqlQuery.append("CONSTRUCT { <qatest1> <qatest2> <qatest3> }");
		sparqlQuery.append(newline);
		sparqlQuery.append("WHERE { ?s ?p ?o . } LIMIT 1");
							
		SPARQLQueryDefinition qdef = sparqlQmgr.newQueryDefinition(sparqlQuery.toString());	
		
		// Base URI set.
		qdef.setBaseUri("http://qa.marklogic.com/functional/tests/one/");
		
		// Verifying Git Issue 356 also, below with JacksonHandle.
		JacksonHandle jh = new JacksonHandle();
		JsonNode jsonResults = sparqlQmgr.executeConstruct(qdef, jh).get();
		String s = jsonResults.fieldNames().next();
		String p = jsonResults.get(s).fieldNames().next();
		JsonNode o = jsonResults.get(s).get(p);
		
		// Verify the mimetype of the handle.
		assertEquals("Method testBaseUri MIME type is incorrect",RDFMimeTypes.RDFJSON,jh.getMimetype());
		assertEquals("Method testBaseUri in-memory subject base URI is incorrect", "http://qa.marklogic.com/functional/tests/one/qatest1", s);
		assertEquals("Method testBaseUri in-memory predicte base URI is incorrect", "http://qa.marklogic.com/functional/tests/one/qatest2", p);
		assertEquals("Method testBaseUri in-memory object base URI is incorrect", "http://qa.marklogic.com/functional/tests/one/qatest3", o.path(0).path("value").asText());	
		
		// Verify if defaulting to RDFJSON when XMLHandle is used. Git Issue 356.
		DOMHandle handle = new DOMHandle();
		Document xmlDoc  = sparqlQmgr.executeConstruct(qdef, handle).get();
		Node description = xmlDoc.getFirstChild().getFirstChild();
		NamedNodeMap attrs = description.getFirstChild().getAttributes();
		
		String oXML = attrs.getNamedItemNS("http://www.w3.org/1999/02/22-rdf-syntax-ns#", "resource").getTextContent();
		
		assertEquals("Method testBaseUri MIME type is incorrect",RDFMimeTypes.RDFXML,handle.getMimetype());
		assertEquals("Method testBaseUri in-memory subject base URI is incorrect", "http://qa.marklogic.com/functional/tests/one/qatest1", description.getAttributes().item(0).getTextContent());
		assertEquals("Method testBaseUri in-memory predicate base URI is incorrect", "qatest2", description.getFirstChild().getNodeName());
		assertEquals("Method testBaseUri in-memory object base URI is incorrect", "http://qa.marklogic.com/functional/tests/one/qatest3", oXML);		
	}

```
